### PR TITLE
HDFS-17312. packetsReceived metric should ignore heartbeat packet.

### PR DESCRIPTION
**Pull request by hfutatzhanghb**
### Description of PR
Metric packetsReceived should ignore heartbeat packet and only used to count data packets and last packet in block.

image

As the above documatation shows, Its original goal is counting total packets received by DataNode excluding heartbeat packet from client. We should fix it.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 30s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  41m 40s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 17s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  compile  |   1m  9s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  checkstyle  |   1m  6s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 21s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  5s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 39s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| -1 :x: |  spotbugs  |   3m 12s | [/branch-spotbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/1/artifact/out/branch-spotbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html) |  hadoop-hdfs-project/hadoop-hdfs in trunk has 1 extant spotbugs warnings.  |
| +1 :green_heart: |  shadedclient  |  34m 25s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javac  |   1m 11s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  3s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 50s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 29s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   3m 10s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  35m  2s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 211m  7s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 42s |  |  The patch does not generate ASF License warnings.  |
|  |   | 345m 27s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6394 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux f0781619503a 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d70974dd98ec356af41f7a21fe74a06ac0c62c54 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/1/testReport/ |
| Max. process+thread count | 3227 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hfutatzhanghb** [comment]
@Hexiaoqiao @zhangshuyan0 @tomscut @tasanuma Sir, could you please review this simple modification when you are free ?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockReceiver.java [hunk 1]
```diff
@@ -598,7 +599,9 @@ private int receivePacket() throws IOException {
       return 0;
     }
 
-    datanode.metrics.incrPacketsReceived();
```
Comment (hotcodemacha) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockReceiver.java:601]: I see you've updated it to increment packetsReceived only when seqno is not HEART_BEAT_SEQNO. This is a critical change for accurately tracking packets. However, it would be helpful to add a comment here explaining why heartbeats are excluded from this metric for clarity and future reference.

**Review by hotcodemacha — CHANGES_REQUESTED** [review]
Can you check if there any other parts of the system that might rely on the current behavior of the packetsReceived metric including heartbeat packets? We need to ensure this change doesn't inadvertently affect other functionalities.

Testing and Validation: Can you confirm if there are existing tests that cover this change or if new tests are needed? It's essential to validate that this modification behaves as expected in various scenarios, including edge cases.

**Comment by hfutatzhanghb** [comment]
> Can you check if there any other parts of the system that might rely on the current behavior of the packetsReceived metric including heartbeat packets? We need to ensure this change doesn't inadvertently affect other functionalities.
> 
> Testing and Validation: Can you confirm if there are existing tests that cover this change or if new tests are needed? It's essential to validate that this modification behaves as expected in various scenarios, including edge cases.

Hi, @ashutoshcipher. Sir. Thanks for your replying. I have checked other parts which using packetsReceived metrics. There are no parts or functionalities rely on the current behavior of the packetsReceived metric including heartbeat packets.  We use find in path to search packetsReceived in all hadoop codes.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockReceiver.java (hunk 1 above)
Comment (hfutatzhanghb) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockReceiver.java:601]: @ashutoshcipher Sir, thanks for your advise. The documentation of packetsReceived metric is showed as below:
image

Its original goal is counting total packets received by DataNode excluding heartbeat packet from client.

**Comment by hfutatzhanghb** [comment]
@Hexiaoqiao @zhangshuyan0 @ashutoshcipher Sir, have updated this PR's description. Please take a look

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 30s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  41m 25s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 17s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  compile  |   1m 12s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  checkstyle  |   1m  6s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 18s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  6s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| -1 :x: |  spotbugs  |   3m 10s | [/branch-spotbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/2/artifact/out/branch-spotbugs-hadoop-hdfs-project_hadoop-hdfs-warnings.html) |  hadoop-hdfs-project/hadoop-hdfs in trunk has 1 extant spotbugs warnings.  |
| +1 :green_heart: |  shadedclient  |  34m 33s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  7s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 10s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javac  |   1m 10s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |   1m  4s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 10s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 51s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   1m 29s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   3m 13s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  34m 10s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 212m 56s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 42s |  |  The patch does not generate ASF License warnings.  |
|  |   | 346m 17s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6394 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 4156fe6a1694 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 9ef6f7216a4781d69cf5a2d7c1543179b7d5b7f3 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/2/testReport/ |
| Max. process+thread count | 5099 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6394/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by tasanuma — APPROVED** [review]
Thanks for reporting the issue. I think the change is appropriate. I also agree that there is no other parts that relies on the current code. +1.

**Comment by tasanuma** [comment]
The spotbugs warning is not related to this PR.

**Comment by tasanuma** [comment]
Thanks for your contribution, @hfutatzhanghb. Thanks for your comments, @ashutoshcipher.

**Comment by hfutatzhanghb** [comment]
@tasanuma Sir, thanks a lot for your reviewing and merging~  @ashutoshcipher Thanks for your reviewing~

**Comment by hfutatzhanghb** [comment]
> Thanks for your contribution, @hfutatzhanghb. Thanks for your comments, @ashutoshcipher.

@tasanuma  Hi, sir. BTW, could you please add ‘Contributed by xxx’ in the commit message . hahaha~

**Comment by tasanuma** [comment]
I believe the author of the commit, that can be confirmed by git log, is undoubtedly the contributor of the commit, so I haven't been including names in the commit messages. But if you prefer, I'll include your name from the next time. :)

**Comment by hfutatzhanghb** [comment]
> I believe the author of the commit, that can be confirmed by git log, is undoubtedly the contributor of the commit, so I haven't been including names in the commit messages. But if you prefer, I'll include your name from the next time. :)

HaHa, Sir, thanks a lot for replying, It's all Okay for you whether including or not including nick name in commit message. I just like the nick "farmmamba"~  Thanks again~
